File: src/main/java/com/chatchatabc/jpademojava/impl/domain/service/UserServiceImpl.java
package com.chatchatabc.jpademojava.impl.domain.service;

import com.chatchatabc.jpademojava.domain.enums.ROLE_NAMES;
import com.chatchatabc.jpademojava.domain.model.Role;
import com.chatchatabc.jpademojava.domain.model.User;
import com.chatchatabc.jpademojava.domain.repository.RoleRepository;
import com.chatchatabc.jpademojava.domain.repository.UserRepository;
import com.chatchatabc.jpademojava.domain.service.UserService;
import org.modelmapper.ModelMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Optional;
import java.util.Set;

@Service
public class UserServiceImpl implements UserService {
    @Autowired
    private UserRepository userRepository;
    @Autowired
    private PasswordEncoder passwordEncoder;
    @Autowired
    private RoleRepository roleRepository;

    private final ModelMapper modelMapper = new ModelMapper();


    /**
     * Register new user
     *
     * @param user
     * @return
     */
    @Override
    public User register(User user) {
        // Encrypt password
        User newUser = modelMapper.map(user, User.class);
        newUser.setPassword(passwordEncoder.encode(newUser.getPassword()));

        // TODO: This can be moved to a separate service
        // Assign Role to user
        long count = userRepository.count();
        // Admin if first registered
        if (count == 0) {
            Optional<Role> adminRole = roleRepository.findRoleByName(ROLE_NAMES.ROLE_ADMIN.toString());
            if (adminRole.isEmpty()) {
                throw new RuntimeException("Admin role not found");
            }
            newUser.setRoles(Set.of(adminRole.get()));
        } else {
            Optional<Role> userRole = roleRepository.findRoleByName(ROLE_NAMES.ROLE_USER.toString());
            if (userRole.isEmpty()) {
                throw new RuntimeException("User role not found");
            }
            newUser.setRoles(Set.of(userRole.get()));
        }
        return userRepository.save(newUser);
    }

    /**
     * Update user profile
     *
     * @param userId
     * @param newUserInfo
     * @return
     */
    @Transactional
    @Override
    public User update(String userId, User newUserInfo) throws Exception {
        Optional<User> queriedUser = userRepository.findById(userId);
        if (queriedUser.isEmpty()) {
            throw new Exception("User not found");
        }

        // Apply update to fields if not null
        // TODO: Could be improved?
        if (newUserInfo.getEmail() != null) {
            queriedUser.get().setEmail(newUserInfo.getEmail());
        }
        if (newUserInfo.getUsername() != null) {
            queriedUser.get().setUsername(newUserInfo.getUsername());
        }
        return userRepository.save(queriedUser.get());
    }

    /**
     * Update user password
     *
     * @param userId
     * @param oldPassword
     * @param newPassword
     * @return
     */
    @Transactional
    @Override
    public User updatePassword(String userId, String oldPassword, String newPassword) throws Exception {
        Optional<User> user = userRepository.findById(userId);
        if (user.isEmpty()) {
            throw new Exception("User not found");
        }
        // Compare old password with current password
        if (!passwordEncoder.matches(oldPassword, user.get().getPassword())) {
            throw new Exception("Old password is incorrect");
        }
        // Update password
        user.get().setPassword(passwordEncoder.encode(newPassword));
        return userRepository.save(user.get());
    }

    /**
     * Load by username for Spring Security
     *
     * @param username
     * @return
     * @throws UsernameNotFoundException
     */
    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        Optional<User> user = userRepository.findByUsername(username);
        if (user.isEmpty()) {
            throw new UsernameNotFoundException("User not found");
        }
        return new org.springframework.security.core.userdetails.User(
                user.get().getUsername(),
                user.get().getPassword(),
                user.get().isEnabled(),
                user.get().isAccountNonExpired(),
                user.get().isCredentialsNonExpired(),
                user.get().isAccountNonLocked(),
                user.get().getAuthorities()
        );
    }
}

File: src/main/kotlin/com/chatchatabc/jpademo/impl/domain/service/UserServiceImpl.kt
package com.chatchatabc.jpademo.impl.domain.service

import com.chatchatabc.jpademo.application.dto.user.UserProfileUpdateRequest
import com.chatchatabc.jpademo.application.dto.user.UserRegisterRequest
import com.chatchatabc.jpademo.domain.model.User
import com.chatchatabc.jpademo.domain.repository.UserRepository
import com.chatchatabc.jpademo.domain.service.UserService
import org.modelmapper.ModelMapper
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import java.util.*

@Service
class UserServiceImpl(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
) : UserService {
    private val mapper = ModelMapper()

    /**
     * User Register
     */
    override fun register(user: UserRegisterRequest): User {
        // Encrypt password
        val newUser = mapper.map(user, User::class.java)
        newUser.password = passwordEncoder.encode(user.password)
        return userRepository.save(newUser)
    }

    /**
     * Update User Profile
     */
    override fun update(userId: String, user: UserProfileUpdateRequest): User {
        val queriedUser = userRepository.findById(userId)
        if (queriedUser.isEmpty) {
            throw Exception("User not found")
        }

        // Apply update
        queriedUser.get().apply {
            // Update email if not null
            email = user.email ?: email
            // Update username if not null
            username = user.username ?: username
        }
        return userRepository.save(queriedUser.get())

    }

    /**
     * User Login
     */
    override fun loadUserByUsername(username: String): UserDetails {
        val user: Optional<User> = userRepository.findByUsername(username)
        if (user.isEmpty) {
            throw Exception("User not found")
        }
        return org.springframework.security.core.userdetails.User(
            user.get().username,
            user.get().password,
            user.get().isEnabled,
            user.get().isAccountNonExpired,
            user.get().isCredentialsNonExpired,
            user.get().isAccountNonLocked,
            user.get().authorities
        )
    }
}